File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/values/PlcLDATE.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.spi.values;

import org.apache.plc4x.java.api.types.PlcValueType;
import org.apache.plc4x.java.spi.codegen.WithOption;
import org.apache.plc4x.java.spi.generation.SerializationException;
import org.apache.plc4x.java.spi.generation.WriteBuffer;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneOffset;

public class PlcLDATE extends PlcIECValue<LocalDate> {

    public static PlcLDATE of(Object value) {
        if (value instanceof PlcLDATE) {
            return (PlcLDATE) value;
        } else if (value instanceof LocalDate) {
            return new PlcLDATE((LocalDate) value);
        } else if (value instanceof Byte) {
            return new PlcLDATE((Byte) value);
        } else if (value instanceof Short) {
            return new PlcLDATE((Short) value);
        } else if (value instanceof Integer) {
            return new PlcLDATE((Integer) value);
        } else if (value instanceof Long) {
            return new PlcLDATE((Long) value);
        } else if (value instanceof Float) {
            return new PlcLDATE((Float) value);
        } else if (value instanceof Double) {
            return new PlcLDATE((Double) value);
        } else if (value instanceof BigInteger) {
            return new PlcLDATE((BigInteger) value);
        } else if (value instanceof BigDecimal) {
            return new PlcLDATE((BigDecimal) value);
        } else {
            return new PlcLDATE(LocalDate.parse(value.toString()));
        }
    }

    public static PlcLDATE ofNanosecondsSinceEpoch(BigInteger nanosecondsSinceEpoch) {
        BigInteger epochSecond = nanosecondsSinceEpoch.divide(BigInteger.valueOf(1000_000));
        BigInteger nanoOfSecond = nanosecondsSinceEpoch.mod(BigInteger.valueOf(1000_000));
        return new PlcLDATE(LocalDateTime.ofEpochSecond(epochSecond.longValue(), nanoOfSecond.intValue(),
            ZoneOffset.UTC).toLocalDate());
    }

    public PlcLDATE(LocalDate value) {
        this.value = value;
        this.isNullable = false;
    }

    public PlcLDATE(Byte secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond((long) secondsSinceEpoch, 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(Short secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond((long) secondsSinceEpoch, 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(Integer secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond((long) secondsSinceEpoch, 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(Long secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond((long) secondsSinceEpoch, 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(Float secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond(secondsSinceEpoch.longValue(), 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(Double secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond(secondsSinceEpoch.longValue(), 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(BigInteger secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond(secondsSinceEpoch.longValue(), 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    public PlcLDATE(BigDecimal secondsSinceEpoch) {
        // REMARK: Yes, I'm using LocalDataTime.ofInstant as LocalDate.ofInstant is marked "JDK 1.9"
        this.value = LocalDateTime.ofEpochSecond(secondsSinceEpoch.longValue(), 0,
            ZoneOffset.UTC).toLocalDate();
        this.isNullable = false;
    }

    @Override
    public PlcValueType getPlcValueType() {
        return PlcValueType.DATE;
    }

    public BigInteger getNanosecondsSinceEpoch() {
        Instant instant = getDateTime().toInstant(ZoneOffset.UTC);
        return BigInteger.valueOf(instant.getEpochSecond()).multiply(BigInteger.valueOf(1000_000_000)).add(BigInteger.valueOf(instant.getNano()));
    }

    @Override
    public boolean isLong() {
        return true;
    }

    @Override
    public long getLong() {
        Instant instant = value.atStartOfDay(ZoneOffset.UTC).toInstant();
        return (instant.toEpochMilli() / 1000);
    }

    @Override
    public boolean isString() {
        return true;
    }

    @Override
    public String getString() {
        return value.toString();
    }

    @Override
    public boolean isDate() {
        return true;
    }

    @Override
    public LocalDate getDate() {
        return value;
    }

    @Override
    public LocalDateTime getDateTime() {
        return value.atStartOfDay();
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }

    @Override
    public void serialize(WriteBuffer writeBuffer) throws SerializationException {
        String valueString = value.toString();
        writeBuffer.writeString(getClass().getSimpleName(),
            valueString.getBytes(StandardCharsets.UTF_8).length * 8,
            valueString, WithOption.WithEncoding(StandardCharsets.UTF_8.name()));
    }

}

File: plc4go/spi/values/PlcLDATE.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package values

import (
	"context"
	"encoding/binary"
	"fmt"
	"time"

	apiValues "github.com/apache/plc4x/plc4go/pkg/api/values"
	"github.com/apache/plc4x/plc4go/spi/utils"
)

type PlcLDATE struct {
	PlcValueAdapter
	value time.Time
}

func NewPlcLDATE(value any) PlcLDATE {
	var timeValue time.Time
	switch value.(type) {
	case time.Time:
		timeValue = value.(time.Time)
	case uint64:
		// Interpreted as "nanoseconds since epoch"
		castedValue := value.(uint64)
		seconds := castedValue / 1000000
		nanoseconds := castedValue % 1000000
		timeValue = time.Unix(int64(seconds), int64(nanoseconds))
	}
	safeValue := time.Date(timeValue.Year(), timeValue.Month(), timeValue.Day(), 0, 0, 0, 0, timeValue.Location())
	return PlcLDATE{
		value: safeValue,
	}
}

func NewPlcLDATEFromNanosecondsSinceEpoch(nanosecondsSinceEpoch uint64) PlcDATE {
	return NewPlcDATE(nanosecondsSinceEpoch)
}

func (m PlcLDATE) GetRaw() []byte {
	theBytes, _ := m.Serialize()
	return theBytes
}

func (m PlcLDATE) GetNanosecondsSinceEpoch() uint64 {
	return uint64(m.value.UnixNano())
}

func (m PlcLDATE) IsDate() bool {
	return true
}

func (m PlcLDATE) GetDate() time.Time {
	return time.Date(m.value.Year(), m.value.Month(), m.value.Day(), 0, 0, 0, 0, time.UTC)
}

func (m PlcLDATE) GetString() string {
	return m.GetDate().Format("2006-01-02")
}

func (m PlcLDATE) GetPlcValueType() apiValues.PlcValueType {
	return apiValues.DATE
}

func (m PlcLDATE) Serialize() ([]byte, error) {
	wb := utils.NewWriteBufferByteBased(utils.WithByteOrderForByteBasedBuffer(binary.BigEndian))
	if err := m.SerializeWithWriteBuffer(context.Background(), wb); err != nil {
		return nil, err
	}
	return wb.GetBytes(), nil
}

func (m PlcLDATE) SerializeWithWriteBuffer(ctx context.Context, writeBuffer utils.WriteBuffer) error {
	return writeBuffer.WriteString("PlcLDATE", uint32(len([]rune(m.GetString()))*8), "UTF-8", m.GetString())
}

func (m PlcLDATE) String() string {
	return fmt.Sprintf("%s(%dbit):%v", m.GetPlcValueType(), uint32(len([]rune(m.GetString()))*8), m.value)
}
